architecture/2025-03-upgrade-design-doc.md

```diff
@@ -0,0 +1,793 @@
+
+    EnableUserDataBackup  bool // Whether automatic user data backup is enabled
+    BackupID              string // ID of user data backup to use for recovery
+}
+```
+
+**Component Health Verification:**
+
+After upgrades, the system verifies all components are healthy:
+
+```go
+type HealthChecker interface {
+    CheckComponentHealth(ctx context.Context, component string) (bool, error)
+    CheckAllComponents(ctx context.Context) (map[string]ComponentHealth, error)
+    WaitForHealthyState(ctx context.Context, timeout time.Duration) error
+}
+
+type ComponentHealth struct {
+    Status      HealthStatus
+    Message     string
+    ReadinessDetails map[string]string
+    LastChecked time.Time
+}
+
+type HealthStatus string
+
+const (
+    StatusHealthy     HealthStatus = "Healthy"
+    StatusDegraded    HealthStatus = "Degraded"
+    StatusUnavailable HealthStatus = "Unavailable"
+    StatusUnknown     HealthStatus = "Unknown"
+)
+```
+
+#### Advantages of this approach
+
+1. **User Experience**: Provides a single command for upgrading all Radius components, significantly simplifying the process compared to manual uninstall/reinstall
+1. **Safety**: Built-in user data backup and restore capabilities ensure user data is protected during upgrades
+1. **Flexibility**: Support for custom configuration parameters allows adaptation to different environments
+1. **Transparency**: Clear, step-by-step output keeps users informed of the upgrade process
+1. **Consistency**: Ensures all Radius components are upgraded together to compatible versions
+
+#### Disadvantages of this approach
+
+1. **Additional Complexity**: Implementing backup/restore functionality adds complexity to the codebase
+1. **Limited Control**: Users have less granular control compared to manually upgrading components
+1. **Resource Requirements**: The upgrade process temporarily requires additional resources during the transition period
+1. **Upgrade Path Constraints**: Some version combinations may require intermediate upgrades, limiting direct jump capability
+1. **CLI Version Mismatch**: Potential confusion for users when their CLI version doesn't match the server version
+
+#### Proposed Option
+
+I recommend implementing the helm-based upgrade approach with automatic user data backup/restore functionality (we can even discuss if we would like this functionality in version 1 or not) because:
+
+1. It provides the best balance between user experience simplicity and technical safety
+2. Leverages existing Helm functionality while adding Radius-specific safety features
+3. The automated user data backup/restore mechanism offers critical protection against data loss
+4. Clear progress indicators and health checks give users confidence in the process
+5. This approach aligns with how users currently install/reinstall Radius, providing consistency
+
+### API design (if applicable)
+
+No specific REST API addition is necessary.
```

Comment:
https://github.com/radius-project/design-notes/blob/9cd00ecbd2898641e68055ad57c01337a680b109/architecture/2025-03-upgrade-design-doc.md?plain=1#L462-L468

I gathered that these will be new APIs, did I misunderstand it?